resources/knowledgebase/tla-review-guidelines.md

```diff
@@ -0,0 +1,147 @@
+
+### Init
+
+* The `Init` predicate in the main spec should be kept general and abstract, representing all valid initial states of the system. For model checking with TLC, use an MC.tla file to specialize the initial state (see Constants and Assumptions section).
+
+### Next
+
+* (Explicit) Actions that leave all variables unchanged are not needed; TLA+ specs are always invariant under stuttering because the next-state relation is written as [Next]_vars, which is equivalent to Next \/ UNCHANGED vars. See `tla-stuttering.md` for details.
+
+* When using action constraints (via TLC's `ACTION_CONSTRAINT` configuration), verify that the constraints make sense given the initial conditions. For example, if `Init` starts with an empty buffer, and only `Produce` is enabled when the buffer is empty, ensure action constraints don't inadvertently prevent the `Produce` action from being explored.
+
+* Be aware of the semantic differences between using if-then-else (ITE) in action definitions versus using enablement conditions. Using ITE (e.g., `IF Len(buffer) < BufferSize THEN ... ELSE UNCHANGED vars`) causes `ENABLED Produce` to always hold (the action is always enabled), whereas an enablement condition (e.g., `Len(buffer) < BufferSize /\ ...`) means the action is only enabled when the buffer has space. This distinction can lead to subtle differences when reasoning about fairness constraints. Choose the appropriate form based on whether the action should be considered enabled in all states or only when specific conditions hold.
+
+### Spec
+
+* In order to check liveness properties, a spec has to assert suitable fairness constraints. In most cases, a fairness constraint asserts weak or strong fairness over the next-state relation, or over individual sub-actions of the next-state relation. Other fairness constraints may result in the spec being non-machine closed, which is a very advanced topic and should be avoided by beginners.
+
+* Fairness constraints should only be added where logically necessary to capture the intended system behavior. Not every action needs fairness - for example, producers that aren't required to produce forever, or actions modeling failures like dropping an item. Consider carefully whether each fairness constraint is justified by the system's actual requirements. Prefer weak fairness (`WF_vars(Consume)`) over strong fairness (`SF_vars(Consume)`) when sufficient, as weak fairness is simpler to reason about and generally adequate unless an action needs to be taken even when only occasionally enabled. Test with TLC to verify whether weak fairness suffices before using strong fairness.
+
+### Invariants
+
+* Every specification should include a `TypeOK` invariant that checks type correctness of all variables. This helps catch type errors early and serves as documentation of the expected types.
```

Comment:
It is good practice to define the `TypeOK` predicate close to the variable declarations so that the reader understands the expected types.
Perhaps we should also say that `TypeOK` is a conjunction of formulas `var \in S` for each variable and an appropriate set S. For example, write `salary \in [Employee -> Nat]` rather than `\A e \in Employee : salary[e] \in Nat`.